hive/src/main/java/org/apache/iceberg/hive/legacy/HiveExpressions.java

```diff
@@ -0,0 +1,271 @@
+   * 5. Removes any children TRUE and FALSE expressions (Note that the simplified expression still can be TRUE and FALSE
+   *    at the root)
+   */
+  private static class Simplify extends ExpressionVisitors.ExpressionVisitor<Expression> {
+
+    private final Set<String> partitionColumnNamesLowerCase;
+
+    Simplify(Set<String> partitionColumnNames) {
+      this.partitionColumnNamesLowerCase =
+          partitionColumnNames.stream().map(String::toLowerCase).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Expression alwaysTrue() {
+      return Expressions.alwaysTrue();
+    }
+
+    @Override
+    public Expression alwaysFalse() {
+      return Expressions.alwaysFalse();
+    }
+
+    @Override
+    public Expression not(Expression result) {
+      return (result == null) ? null : result.negate();
+    }
+
+    @Override
+    public Expression and(Expression leftResult, Expression rightResult) {
+      if (leftResult == null && rightResult == null) {
+        return null;
+      } else if (leftResult == null) {
+        return rightResult;
+      } else if (rightResult == null) {
+        return leftResult;
+      } else if (leftResult.op() == Expression.Operation.FALSE || rightResult.op() == Expression.Operation.FALSE) {
+        return Expressions.alwaysFalse();
+      } else if (leftResult.op() == Expression.Operation.TRUE) {
+        return rightResult;
+      } else if (rightResult.op() == Expression.Operation.TRUE) {
+        return leftResult;
+      } else {
+        return Expressions.and(leftResult, rightResult);
+      }
+    }
+
+    @Override
+    public Expression or(Expression leftResult, Expression rightResult) {
+      if (leftResult == null && rightResult == null) {
+        return null;
+      } else if (leftResult == null || rightResult == null) {
+        throw new IllegalStateException(
+            "A filter on a partition column was ORed with a filter on a non-partition column which is not supported" +
+                " yet");
+      } else if (leftResult.op() == Expression.Operation.TRUE || rightResult.op() == Expression.Operation.TRUE) {
+        return Expressions.alwaysTrue();
+      } else if (leftResult.op() == Expression.Operation.FALSE) {
+        return rightResult;
+      } else if (rightResult.op() == Expression.Operation.FALSE) {
+        return leftResult;
+      } else {
+        return Expressions.or(leftResult, rightResult);
+      }
+    }
+
+    <T> Expression in(UnboundTerm<T> expr, List<Literal<T>> literals) {
+      Expression in = alwaysFalse();
+      for (Literal<T> literal : literals) {
+        in = Expressions.or(in, Expressions.equal(expr, literal.value()));
+      }
+      return in;
+    }
+
+    @Override
+    public <T> Expression predicate(BoundPredicate<T> pred) {
+      throw new IllegalStateException("Bound predicate not expected: " + pred.getClass().getName());
```

Comment:
Rephrase?

Reply:
I think this is fine. This is not a user facing message. User should never see this message if we write our code correctly.